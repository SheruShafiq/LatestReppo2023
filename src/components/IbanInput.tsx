import { TextField } from "@mui/material";
import { validateIBAN } from "@/lib/helper/Regex";
import React from "react";

export type IbanInputProps = {
  value: string;
  change: (event: React.ChangeEvent<HTMLInputElement>) => void;
};
const IbanInput = ({ value, change }: IbanInputProps) => {
  const [error, setError] = React.useState(false);

  function ibanValidation(iban: string) {
    if (iban.length === 18) {
      setError(!validateIBAN.test(iban));
    } else {
      setError(true);
    }
  }

  return (
    <TextField
      label="IBAN"
      variant="outlined"
      value={value}
      onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
        change(event);
      }}
      onBlur={() => ibanValidation(value)}
      sx={{ height: "2.5rem" }}
      size="small"
      error={error}
    />
  );
};

export default IbanInput;
