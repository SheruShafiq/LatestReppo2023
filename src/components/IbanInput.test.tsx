// Import dependencies
import { render, fireEvent } from "@testing-library/react";
import IbanInput from "./IbanInput";
import matchers from "@testing-library/jest-dom";
import { debug } from "vitest-preview";
expect.extend(matchers);

describe("IbanInput", () => {
  const changeMock = vi.fn();

  beforeEach(() => {
    vi.resetAllMocks();
  });

  it("renders successfully", () => {
    const { getByLabelText } = render(
      <IbanInput value="Test Iban" change={changeMock} />
    );
    expect(getByLabelText("IBAN")).toBeInTheDocument();
  });

  it("should call the passed change function on input change", async () => {
    const { getByLabelText } = render(
      <IbanInput value="Test Iban" change={changeMock} />
    );
    const ibanInput = getByLabelText("IBAN") as HTMLInputElement;
    await fireEvent.input(ibanInput, { target: { value: "New Test Iban" } });
    expect(changeMock).toHaveBeenCalled();
  });

  it("should validate the iban on blur", async () => {
    const { getByLabelText } = render(
      <IbanInput value="NL68INGB06709910dd588" change={changeMock} />
    );

    const ibanInput = getByLabelText("IBAN") as HTMLInputElement;

    const label = getByLabelText("IBAN").parentElement;

    await fireEvent.blur(ibanInput);
    debug();
    expect(label).toHaveClass("Mui-error");
  });

  it("should not set error for valid IBAN", async () => {
    // Here, input a valid IBAN string
    const { getByLabelText } = render(
      <IbanInput value="NL68INGB0670991058" change={changeMock} />
    );
    const ibanInput = getByLabelText("IBAN") as HTMLInputElement;

    // Trigger the onBlur event
    await fireEvent.blur(ibanInput);

    // Expect the input to not have error
    const label = getByLabelText("IBAN").parentElement;

    expect(label).not.toHaveClass("Mui-error");
    expect(screen).toMatchSnapshot();
  });
});
