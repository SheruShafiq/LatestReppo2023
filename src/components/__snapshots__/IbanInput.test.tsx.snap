// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IbanInput > should not set error for valid IBAN 1`] = `Screen {}`;
